### PR TITLE
1189 fetch diff

### PR DESCRIPTION
**Pull request by elizabethengelman**
companion mcd PR: https://github.com/makerdao/vdb-mcd-transformers/pull/124

### CODE REVIEW

File: libraries/shared/transformer/storage_transformer.go [hunk 1]
```diff
@@ -25,6 +25,8 @@ import (
 type StorageTransformer interface {
 	Execute(diff types.PersistedDiff) error
 	KeccakContractAddress() common.Hash
+	GetStorageKeysLookup() interface{}
```
Comment (elizabethengelman) [inline libraries/shared/transformer/storage_transformer.go]: I'd love some feedback on this. Because of the way the project is structured, if we have this return a `storage.KeysLookup` we end up having a circular dependency because the `storage` package imports this `transformer` package. 

I'm wondering if it makes sense to have this `StorageTransformer` interface here in it's own package, or if we could potentially move it to the `storage` package where the concrete Transformer is being implemented. I'm thinking of a similar change for the event transformer interface. 🤔

File: cmd/getStorageValue.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+	blockNumber     int64
+}
+
+func (r *StorageValueCommandRunner) Run() error {
+	addressToKeys, getKeysErr := r.getStorageKeys()
+	if getKeysErr != nil {
+		return getKeysErr
+	}
+
+	header, getHeaderErr := r.headerRepo.GetHeader(r.blockNumber)
```
Comment (elizabethengelman) [inline cmd/getStorageValue.go]: Currently this returns an error if we don't have a header for the given block in the db yet. I _think_ that should be okay, because we'll likely not run this command super close to the head of the chain, but I'd be curious to hear what folks think about that. The other idea I had was to get the header (which we need to get the hash for the storage_diff table) directly from the blockchain, but thought that saving ourselves another rpc call may be worth it, especially since we already should have the header in the db.

File: cmd/getStorageValue.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+		}
+	}
+
+	return nil
+}
+
+func (r *StorageValueCommandRunner) getAndPersistStorageValues(address common.Address, keys []common.Hash, blockNumber int64, headerHash string) error {
+	blockNumberBigInt := big.NewInt(blockNumber)
+	for _, key := range keys {
+		value, getStorageErr := r.bc.GetStorageAt(address, key, blockNumberBigInt)
```
Comment (elizabethengelman) [inline cmd/getStorageValue.go]: Thinking that a future performance boost could be to batch these `GetStorageAt` calls.

File: db/migrations/00005_create_storage_diffs_table.sql [hunk 4]
```diff
@@ -8,6 +8,7 @@ CREATE TABLE public.storage_diff
     storage_key    BYTEA,
     storage_value  BYTEA,
     checked        BOOLEAN NOT NULL DEFAULT FALSE,
+    from_backfill  BOOLEAN NOT NULL DEFAULT FALSE,
```
Comment (elizabethengelman) [inline db/migrations/00005_create_storage_diffs_table.sql]: I'd be happy to remove this change until after the release if we don't want to change the schema this last minute.

File: libraries/shared/factories/storage/transformer.go [hunk 5]
```diff
@@ -25,11 +25,20 @@ import (
 )
 
 type Transformer struct {
+	Address           common.Address
```
Comment (rmulhol) [inline libraries/shared/factories/storage/transformer.go]: Double checking - is it accurate that we'll want a corresponding PR to update all of the initializers for the storage transformers to include this field?

File: libraries/shared/mocks/storage_diff_repository.go [hunk 6]
```diff
@@ -47,3 +49,17 @@ func (repository *MockStorageDiffRepository) CreateStorageDiff(rawDiff types.Raw
 	repository.CreatePassedRawDiffs = append(repository.CreatePassedRawDiffs, rawDiff)
 	return repository.CreateReturnID, repository.CreateReturnError
 }
+
+func (repository *MockStorageDiffRepository) SetCreateError(err error) {
```
Comment (rmulhol) [inline libraries/shared/mocks/storage_diff_repository.go]: Just curious - what led you to opt for creating a setter here, as opposed to writing to the property directly?

File: libraries/shared/mocks/storage_repository.go [hunk 7]
```diff
@@ -37,6 +38,6 @@ func (repository *MockStorageRepository) Create(diffID, headerID int64, metadata
 	return repository.CreateErr
 }
 
-func (*MockStorageRepository) SetDB(db *postgres.DB) {
-	panic("implement me")
+func (repository *MockStorageRepository) SetDB(db *postgres.DB) {
+	repository.db = db
```
Comment (rmulhol) [inline libraries/shared/mocks/storage_repository.go]: Maybe this should be a no-op? Thinking that we never want the mock to actually interact with the db

File: libraries/shared/transformer/storage_transformer.go (hunk 1 above)
Comment (rmulhol) [reply libraries/shared/transformer/storage_transformer.go]: 💯 to moving it, and returning a more specific type

File: pkg/fakes/mock_blockchain.go [hunk 8]
```diff
@@ -109,6 +115,28 @@ func (blockChain *MockBlockChain) LastBlock() (*big.Int, error) {
...
+	blockChain.GetStorageAtPassedKeys = append(blockChain.GetStorageAtPassedKeys, key)
+	blockChain.GetStorageAtPassedBlockNumber = blockNumber
+
+	if len(blockChain.storageValuesToReturn) > 0 {
+		storageToReturn := blockChain.storageValuesToReturn[account]
+
+		return storageToReturn, blockChain.GetStorageAtError
+	} else {
+		return nil, blockChain.GetStorageAtError
+	}
```
Comment (rmulhol) [inline pkg/fakes/mock_blockchain.go]: Do we need this `if`/`else`? I would have thought that line 126 would return the same thing as line 128 if there was no entry for the account, but could be wrong

File: pkg/fakes/mock_eth_client.go [hunk 9]
```diff
@@ -62,29 +67,7 @@ type MockEthClient struct {
...
-		filterLogsErr:               nil,
-		filterLogsPassedContext:     nil,
-		filterLogsPassedQuery:       ethereum.FilterQuery{},
-		filterLogsReturnLogs:        nil,
-		transactionReceipts:         make(map[string]*types.Receipt),
-		err:                         nil,
-		passedBatch:                 nil,
-		passedMethod:                "123",
-	}
+	return &MockEthClient{}
```
Comment (rmulhol) [inline pkg/fakes/mock_eth_client.go]: 🎉

File: pkg/fakes/mock_eth_client.go [hunk 10]
```diff
@@ -137,6 +120,14 @@ func (client *MockEthClient) SetTransactionSenderErr(err error) {
 	client.transactionSenderErr = err
 }
 
+func (client *MockEthClient) SetBalanceAtErr(err error) {
+	client.balanceAtErr = err
+}
+
+func (client *MockEthClient) SetBalanceAt(balance *big.Int) {
+	client.passedBalance = balance
```
Comment (rmulhol) [inline pkg/fakes/mock_eth_client.go]: maybe this should be called `balanceToReturn` or something? `passedBalance` makes me think it's an argument received by one of the functions

File: cmd/getStorageValue.go [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+// getStorageValueCmd represents the getStorageValue command
+var getStorageValueCmd = &cobra.Command{
+	Use:   "getStorageValue",
+	Short: "Gets all storage values for configured contracts at the given block.",
+	Long: `Fetches and persists storage values of the configured contracts at a given block. It is important to note that the storage value gotten with this
+	command may not be different from the previous block in the database.`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		SubCommand = cmd.CalledAs()
+		LogWithCommand = *logrus.WithField("SubCommand", SubCommand)
+		fmt.Println("getStorageValue called")
```
Comment (rmulhol) [inline cmd/getStorageValue.go]: super nit-picky but maybe it makes sense to replace this with something like `fmt.Printf("getting storage values for all known keys at block %d", getStorageValueBlockNumber)`?

File: cmd/getStorageValue.go [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+		return getStorageAt(getStorageValueBlockNumber)
```
Comment (rmulhol) [inline cmd/getStorageValue.go]: Do you know what happens if this returns an error? Hoping to confirm it communicates that back to the CLI if unsuccessful

File: cmd/getStorageValue.go [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+		bc:              bc,
+		db:              db,
+		HeaderRepo:      repositories.NewHeaderRepository(db),
+		StorageDiffRepo: storage2.NewDiffRepository(db),
+		initializers:    initializers,
+		blockNumber:     blockNumber,
+	}
+}
+
+type StorageValueCommandRunner struct {
```
Comment (rmulhol) [inline cmd/getStorageValue.go]: Why not pull this out into a separate namespace outside of the `cmd` directory? Totally fine with kicking the can down the road if it's a hassle

File: cmd/getStorageValue.go [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+func (r *StorageValueCommandRunner) getStorageKeys() (map[common.Address][]common.Hash, error) {
+	addressToKeys := make(map[common.Address][]common.Hash)
+	for _, i := range r.initializers {
+		transformer := i(r.db)
+		keysLookup, ok := transformer.GetStorageKeysLookup().(storage.KeysLookup)
+		if !ok {
+			errorString := fmt.Sprintf("%v type incompatible. Should be a storage.KeysLookup", keysLookup)
+			return addressToKeys, errors.New(errorString)
```
Comment (rmulhol) [inline cmd/getStorageValue.go]: nbd but could use `fmt.Errorf`

File: cmd/getStorageValue_test.go [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+
+		headerRepo = fakes.MockHeaderRepository{}
+		fakeHeader = fakes.FakeHeader
+		fakeHeader.BlockNumber = blockNumber
+		headerRepo.GetHeaderReturnHash = fakeHeader.Hash
+		runner.HeaderRepo = &headerRepo
+	})
+
+	AfterEach(func() {
+		test_config.CleanTestDB(db)
```
Comment (rmulhol) [inline cmd/getStorageValue_test.go]: Why do this in an `AfterEach` rather than `BeforeEach`?

Also, do these tests actually need a `db`, or could we maybe get away with passing `nil`?

File: db/migrations/00005_create_storage_diffs_table.sql (hunk 4 above)
Comment (rmulhol) [reply db/migrations/00005_create_storage_diffs_table.sql]: Would be up for merging the migrations early/separately too if you think merging this PR will happen after the release

**Comment by rmulhol** [comment]
Have you had a chance to try this out locally? Would be curious to see how it runs/whether we want to add some intermittent logging on the happy path (if it takes awhile)

File: libraries/shared/transformer/storage_transformer.go (hunk 1 above)
Comment (elizabethengelman) [reply libraries/shared/transformer/storage_transformer.go]: 👍 updated in b544c91ff214a787ec509b8ba75fb7eaf2df29e8

will likely do the same for the EventTransformer and ContractTransformer interfaces in another PR.

File: libraries/shared/factories/storage/transformer.go (hunk 5 above)
Comment (elizabethengelman) [reply libraries/shared/factories/storage/transformer.go]: Yep, that's correct. Thanks for the reminder!

File: libraries/shared/mocks/storage_diff_repository.go (hunk 6 above)
Comment (elizabethengelman) [reply libraries/shared/mocks/storage_diff_repository.go]: no idea 🤷‍♀ 😜  i'll remove that Set method in favor of just setting the property directly.

File: pkg/fakes/mock_blockchain.go (hunk 8 above)
Comment (elizabethengelman) [reply pkg/fakes/mock_blockchain.go]: yeah, nice catch! that was some cruft from refactoring the mock GetStorageAt method. ✂️

File: cmd/getStorageValue_test.go (hunk 15 above)
Comment (elizabethengelman) [reply cmd/getStorageValue_test.go]: nice catch!

File: libraries/shared/transformer/storage_transformer.go (hunk 1 above)
Comment (elizabethengelman) [reply libraries/shared/transformer/storage_transformer.go]: Actually, I am going to remove that change from this PR - as I was testing this out, I realized that it will affect the way that the [plugin is built ](https://github.com/makerdao/vulcanizedb/blob/staging/pkg/plugin/writer/writer.go#L60). So I think I'll hold off on that so that this PR isn't too big, and create a story for dealing with this for storage, event and contract transformers in the future.

File: cmd/getStorageValue.go (hunk 13 above)
Comment (elizabethengelman) [reply cmd/getStorageValue.go]: I like that idea! Updated in 1e72fcc9415a49119be9447cc1d46558137508f3, I'd be interested in thoughts about the naming.

File: cmd/getStorageValue.go (hunk 12 above)
Comment (elizabethengelman) [reply cmd/getStorageValue.go]: Yep, it returns an error and then returns the command's usage dialog:

```
elizabethengelman@eme:vulcanizedb (1189-fetch-diff) 💾  $ ./vulcanizedb getStorageValue --config ../vdb-mcd-transformers/environments/mcdTransformers.toml
Error: sql: no rows in result set
Usage:
  vulcanizedb getStorageValue [flags]

Flags:
  -b, --get-storage-value-block-number int   block number to fetch storage at for all configured transformers (default -1)
  -h, --help                                 help for getStorageValue

Global Flags:
      --client-ipcPath string                location of geth.ipc file
...
```

It also includes a fatal log in vulcanizedb.log.

**Comment by elizabethengelman** [comment]
I have tried it out locally, but not with a large amount of data - but I like the idea of more logging, just pushed up a change. I'm wondering if we'd want even more logging, like every 10th diff value or something?

File: cmd/getStorageValue.go [hunk 16]
```diff
@@ -0,0 +1,67 @@
+// Copyright © 2020 NAME HERE <EMAIL ADDRESS>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (rmulhol) [inline cmd/getStorageValue.go]: Should probably remove the Apache license

File: cmd/root.go [hunk 17]
```diff
@@ -220,3 +222,45 @@ func prepConfig() error {
 	}
 	return nil
 }
+
+func exportTransformers() []transformer.StorageTransformerInitializer {
```
Comment (rmulhol) [inline cmd/root.go]: Maybe it would make sense to name this `exportStorageTransformers`?

File: cmd/getStorageValue.go [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+
+func init() {
+	getStorageValueCmd.Flags().Int64VarP(&getStorageValueBlockNumber, getStorageFlagName, "b", -1, "block number to fetch storage at for all configured transformers")
+	rootCmd.AddCommand(getStorageValueCmd)
+}
+
+func getStorageAt(blockNumber int64) error {
+	blockChain := getBlockChain()
+	db := utils.LoadPostgres(databaseConfig, blockChain.Node())
+	storageInitializers := exportTransformers()
```
Comment (rmulhol) [inline cmd/getStorageValue.go]: Does it make sense to potentially do a nil check in case there are no storage initializers in the config?

File: cmd/getStorageValue.go [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+		LogWithCommand.Infof("Getting storage values for all known keys at block %d", getStorageValueBlockNumber)
+
+		validationErr := validateBlockNumberArg(getStorageValueBlockNumber, getStorageFlagName)
+		if validationErr != nil {
+			return validationErr
+		}
+
+		getStorageErr := getStorageAt(getStorageValueBlockNumber)
+		if getStorageErr != nil {
+			return fmt.Errorf("SubCommand %v: Failed to get storage values at block %v. Err: %v", SubCommand, getStorageValueBlockNumber, getStorageErr)
```
Comment (rmulhol) [inline cmd/getStorageValue.go]: nbd but I think the linter complains about capitalization and punctuation (except `:`) in errors

File: cmd/resetHeaderCheckCount.go [hunk 20]
```diff
@@ -26,49 +25,42 @@ import (
...
-		validationErr := validateBlockNumberArg(blockNumber)
+		validationErr := validateBlockNumberArg(resetHeaderCountBlockNumber, resetHeaderFlagName)
 		if validationErr != nil {
-			errorString := fmt.Sprintf("%v: header-block-number argument is required and not value was given.", SubCommand)
-			return errors.New(errorString)
+			return validationErr
 		}
 
-		resetErr := resetHeaderCount(int64(blockNumber))
+		resetErr := resetHeaderCount(int64(resetHeaderCountBlockNumber))
```
Comment (rmulhol) [inline cmd/resetHeaderCheckCount.go]: Might not need to cast this to an `int64`?

File: cmd/root.go [hunk 21]
```diff
@@ -220,3 +222,45 @@ func prepConfig() error {
...
+	LogWithCommand.Info("loading transformers from plugin")
+	symExporter, lookupErr := plug.Lookup("Exporter")
+	if lookupErr != nil {
+		LogWithCommand.Fatalf("loading Exporter symbol failed: %s", lookupErr.Error())
+	}
+
+	// Assert that the symbol is of type Exporter
+	exporter, ok := symExporter.(Exporter)
+	if !ok {
+		LogWithCommand.Fatal("plugged-in symbol not of type Exporter")
```
Comment (rmulhol) [inline cmd/root.go]: nbd but I'd be a fan of returning an error and going the `Fatal` logging from the caller - don't think that has a material consequence here but prob a good habit in case there are surrounding processes that need to shutdown gracefully

File: libraries/shared/mocks/storage_keys_loader.go [hunk 22]
```diff
@@ -37,3 +37,11 @@ func (loader *MockStorageKeysLoader) LoadMappings() (map[common.Hash]types.Value
 func (loader *MockStorageKeysLoader) SetDB(db *postgres.DB) {
 	loader.SetDBCalled = true
 }
+
+func (loader *MockStorageKeysLoader) SetMappingsToLoad(mappings map[common.Hash]types.ValueMetadata) {
+	loader.StorageKeyMappings = mappings
+}
+
+func (loader *MockStorageKeysLoader) SetLoadMappingError(err error) {
```
Comment (rmulhol) [inline libraries/shared/mocks/storage_keys_loader.go]: Maybe don't need these setters?

File: libraries/shared/mocks/storage_keys_lookup.go [hunk 23]
```diff
@@ -23,16 +23,33 @@ import (
...
 
-func (*MockStorageKeysLookup) SetDB(db *postgres.DB) {
-	panic("implement me")
+func (lookup *MockStorageKeysLookup) GetKeys() ([]common.Hash, error) {
+	lookup.GetKeysCalled = true
+	return lookup.keysToReturn, lookup.getKeysError
+}
+
+func (lookup *MockStorageKeysLookup) SetDB(db *postgres.DB) {
+	lookup.db = db
```
Comment (rmulhol) [inline libraries/shared/mocks/storage_keys_lookup.go]: Should this be a no-op?

File: libraries/shared/mocks/storage_keys_lookup.go [hunk 24]
```diff
@@ -23,16 +23,33 @@ import (
...
+}
+
+func (lookup *MockStorageKeysLookup) SetKeysToReturn(keys []common.Hash) {
+	lookup.keysToReturn = keys
+}
+
+func (lookup *MockStorageKeysLookup) SetGetKeysError(err error) {
```
Comment (rmulhol) [inline libraries/shared/mocks/storage_keys_lookup.go]: Maybe don't need setters here

File: libraries/shared/storage/backfill/storage_value_loader.go [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+	initializers    []transformer.StorageTransformerInitializer
+	blockNumber     int64
+}
+
+func (r *StorageValueLoader) Run() error {
+	addressToKeys, getKeysErr := r.getStorageKeys()
+	if getKeysErr != nil {
+		return getKeysErr
+	}
+	logrus.Infof("Received storage keys for %v addresses.", len(addressToKeys))
```
Comment (rmulhol) [inline libraries/shared/storage/backfill/storage_value_loader.go]: maybe it would make sense to log the addresses for which we're looking up keys?

File: libraries/shared/storage/backfill/storage_value_loader.go [hunk 26]
```diff
@@ -0,0 +1,111 @@
...
+			HashedAddress: crypto.Keccak256Hash(address[:]),
+			BlockHash:     common.HexToHash(headerHash),
+			BlockHeight:   int(blockNumber),
+			StorageKey:    key,
+			StorageValue:  common.BytesToHash(value),
+		}
+
+		diffId, createDiffErr := r.StorageDiffRepo.CreateStorageDiff(diff)
+		if createDiffErr != nil {
+			if createDiffErr == sql.ErrNoRows {
```
Comment (rmulhol) [inline libraries/shared/storage/backfill/storage_value_loader.go]: Does this mean it's a duplicate?

File: libraries/shared/storage/backfill/storage_value_loader.go [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+		}
+	}
+	return nil
+}
+
+func (r *StorageValueLoader) getStorageKeys() (map[common.Address][]common.Hash, error) {
+	addressToKeys := make(map[common.Address][]common.Hash)
+	for _, i := range r.initializers {
+		transformer := i(r.db)
+		keysLookup, ok := transformer.GetStorageKeysLookup().(storage.KeysLookup)
```
Comment (rmulhol) [inline libraries/shared/storage/backfill/storage_value_loader.go]: does this cast potentially cause a nil pointer panic if `!ok`?

**Comment by rmulhol** [comment]
> I have tried it out locally, but not with a large amount of data - but I like the idea of more logging, just pushed up a change. I'm wondering if we'd want even more logging, like every 10th diff value or something?

A little torn about this - thinking we could be looking up 1000s of keys 🤔 
Maybe we could log the number of keys per address?

File: cmd/getStorageValue.go (hunk 16 above)
Comment (elizabethengelman) [reply cmd/getStorageValue.go]: 🤦‍♀ I don't know how that keeps getting in there! My goland copyright profile is empty. 🤔 Thanks for catching that!

File: cmd/root.go (hunk 17 above)
Comment (elizabethengelman) [reply cmd/root.go]: I think I'll keep the same name, and then have it export all of the initializers. Thinking we can reuse this code in the execute command as well

File: cmd/getStorageValue.go (hunk 18 above)
Comment (elizabethengelman) [reply cmd/getStorageValue.go]: good call!

File: libraries/shared/storage/backfill/storage_value_loader.go (hunk 26 above)
Comment (elizabethengelman) [reply libraries/shared/storage/backfill/storage_value_loader.go]: Yep, exactly. The StorageDiffRepo returns a `sql.ErrNoRows` when it encounters a duplicate diff: https://github.com/makerdao/vulcanizedb/blob/staging/libraries/shared/storage/diff_repository.go#L51

File: cmd/root.go (hunk 21 above)
Comment (elizabethengelman) [reply cmd/root.go]: That makes sense to me! I've updated `getStorageValue` command to follow that pattern, but decided to leave `execute` as is for now, and just do a `LogWithCommand.Fatal`. `execute` would require the underlying method to be refactored to return an error, and I am a little apprehensive in dealing with (and it's waitgroups 😱 ) as part of this change since it's already feeling kind of big.

File: cmd/getStorageValue.go (hunk 19 above)
Comment (elizabethengelman) [reply cmd/getStorageValue.go]: Yeah, I think you're right... though I'm not seeing it complain on this one for some reason. 🤔

File: libraries/shared/storage/backfill/storage_value_loader.go (hunk 27 above)
Comment (elizabethengelman) [reply libraries/shared/storage/backfill/storage_value_loader.go]: I think that `ok` is a bool, so by capturing that in a var we'll able to handle it and return an error instead of panicking. Not sure how to test this out in the code itself, but tried it out in a contrived example here: https://play.golang.org/p/UUfNh5tnvLr
